unittests/checktime_tests.cpp

```diff
@@ -150,6 +150,49 @@ BOOST_AUTO_TEST_CASE( checktime_interrupt_test) { try {
 
 } FC_LOG_AND_RETHROW() }
 
+BOOST_AUTO_TEST_CASE( checktime_speculative_max_trx_test ) { try {
+   savanna_tester t;
+   t.produce_block();
+   t.create_account( "pause"_n );
+   t.set_code( "pause"_n, test_contracts::test_api_wasm() );
+   t.produce_block();
+
+   BOOST_CHECK_EXCEPTION( push_trx( t, test_pause_action<WASM_TEST_ACTION("test_checktime", "checktime_failure")>{},
+                                    0, 25, 500, false, fc::raw::pack(10000000000000000000ULL), "pause"_n ),
+                          tx_cpu_usage_exceeded, fc_exception_message_contains("reached node configured max-transaction-time") );
+
+   auto before = fc::time_point::now();
+   // test case where max-transaction-time = -1, UINT32_MAX is converted to fc::microseconds::maximum() in push_trx
+   // Verify restricted to 150ms (on-chain max_transaction_cpu_usage)
+   BOOST_CHECK_EXCEPTION( push_trx( t, test_pause_action<WASM_TEST_ACTION("test_checktime", "checktime_failure")>{},
+                                     100000, UINT32_MAX, 10000, false, fc::raw::pack(10000000000000000000ULL), "pause"_n ),
+                          tx_cpu_usage_exceeded, fc_exception_message_contains("reached on chain max_transaction_cpu_usage 150000us") );
+   auto after = fc::time_point::now();
+   auto dur = (after - before).count();
+   // verify within 20% of expected time
+   BOOST_CHECK_MESSAGE( dur >= 150'000, "elapsed " << dur << "us" );
+   BOOST_CHECK_MESSAGE( dur <= 180'000, "elapsed " << dur << "us" );
```

Comment:
Would it be hard to retrieve the on-chain limit from the controller instead of using a hard coded `150'000` value?